ruby/red-arrow/ext/arrow/record_batch.cc

```diff
@@ -0,0 +1,610 @@
+                                  "Unsupported array in ConvertValue");
+                         return Qnil;
+                       });
+  }
+
+  inline VALUE ConvertValue(const arrow::NullArray& array, const int64_t i) {
+    return Qnil;
+  }
+
+  inline VALUE ConvertValue(const arrow::BooleanArray& array, const int64_t i) {
+    bool value = array.Value(i);
+    return value ? Qtrue : Qfalse;
+  }
+
+  template <typename ArrayType>
+  inline VALUE ConvertSignedIntegerValue(const ArrayType& array, const int64_t i) {
+    return rb::protect([&]{ return LL2NUM(array.Value(i)); });
+  }
+
+#define CONVERT_SIGNED_INTEGER(TYPE) \
+  VALUE ConvertValue(const TYPE& array, const int64_t i) { \
+    return ConvertSignedIntegerValue<TYPE>(array, i); \
+  }
+
+  CONVERT_SIGNED_INTEGER(arrow::Int8Array)
+  CONVERT_SIGNED_INTEGER(arrow::Int16Array)
+  CONVERT_SIGNED_INTEGER(arrow::Int32Array)
+  CONVERT_SIGNED_INTEGER(arrow::Int64Array)
+
+#undef CONVERT_SIGNED_INTEGER
+
+  template <typename ArrayType>
+  inline VALUE ConvertUnsignedIntegerValue(const ArrayType& array, const int64_t i) {
+    return rb::protect([&]{ return ULL2NUM(array.Value(i)); });
+  }
+
+#define CONVERT_UNSIGNED_INTEGER(TYPE) \
+  VALUE ConvertValue(const TYPE& array, const int64_t i) { \
+    return ConvertUnsignedIntegerValue<TYPE>(array, i); \
+  }
+
+  CONVERT_UNSIGNED_INTEGER(arrow::UInt8Array)
+  CONVERT_UNSIGNED_INTEGER(arrow::UInt16Array)
+  CONVERT_UNSIGNED_INTEGER(arrow::UInt32Array)
+  CONVERT_UNSIGNED_INTEGER(arrow::UInt64Array)
+
+#undef CONVERT_UNSIGNED_INTEGER
+
+  template <typename ArrayType>
+  inline VALUE ConvertFloatValue(const ArrayType& array, const int64_t i) {
+    return rb::protect([&]{ return DBL2NUM(array.Value(i)); });
+  }
+
+#define CONVERT_FLOAT(TYPE) \
+  VALUE ConvertValue(const TYPE& array, const int64_t i) { \
+    return ConvertFloatValue<TYPE>(array, i); \
+  }
+
+  CONVERT_FLOAT(arrow::FloatArray)
+  CONVERT_FLOAT(arrow::DoubleArray)
+
+#undef CONVERT_FLOAT
+
+  inline VALUE ConvertValue(const arrow::HalfFloatArray& array, const int64_t i) {
+    // FIXME: should convert to Float
+    return rb::protect([&]{
+      return LONG2FIX(static_cast<long>(array.Value(i)));
+    });
+  }
+
+  inline VALUE ConvertDecimal128(const arrow::Decimal128Array& array, const int64_t i) {
+    return rb::protect([&]{
+      auto raw_value = std::make_shared<arrow::Decimal128>(array.GetValue(i));
+      auto gobj_value = garrow_decimal128_new_raw(&raw_value);
+      return GOBJ2RVAL(gobj_value);
+    });
+  }
+
+  inline VALUE ConvertDecimal128ToBigDecimal(const arrow::Decimal128Array& array, const int64_t i) {
+    auto decimal_str = array.FormatValue(i);
+    return rb::protect([&]{
+      return rb_funcall(rb_cObject, id_BigDecimal, 1, rb_str_new_cstr(decimal_str.c_str()));
+    });
+  }
+
+  inline VALUE ConvertValue(const arrow::Decimal128Array& array, const int64_t i) {
+    if (convert_decimal_) {
+      return ConvertDecimal128ToBigDecimal(array, i);
```

Comment:
I consider this conversion.
We can always convert to `BigDecimal` because this is `raw_records` converter. We can always use standard Ruby objects.

Reply:
I removed `convert_decimal:` optional keyword argument.